src/code-considerations/using-unstable-lang/const-evaluation.md

```diff
@@ -0,0 +1,100 @@
+# Const evaluation
+
+Rust's compile-time code evaluation story is still in an experimental state,
+with a large amount of unstable surface area, including areas where there is
+active design work.
+
+The standard library (as of this writing) uses some amount of hacks to work
+around language limitations for some implementations: this document provides a
+framework for evaluating the cost/benefit analysis of using such workarounds to
+enable adding new `const` functions/traits or `const`-ifying existing APIs.
+
+The overall high-level guidance is to evaluate the tradeoff between enabling
+experimentation in users of the standard library (or discovery of limitations
+to these language features) and the costs to the maintenance and readability of
+the standard library. (This is broad guidance applicable to almost any feature,
+but const generics are particularly notable in terms of their potentially wide
+scope).
+
+These are good areas to consider in such an evaluation. You should think about
+whether the given PR is adding value to outweigh its costs along these
+dimensions, and add to the PR description what your justification for the usage
+is. Consider including future plans for the feature as well: if the feature is
+experimental, does it have a clear path toward stabilization in the current
+form?
+
+## Maintainability
+
+Does making this API const require that code be hand-written which was previously automatic?
+
+Does the new code require duplicating, for example through `#[cfg(bootstrap)]`,
+increasing the risk of diverging between the two implementations (and
+increasing work when bumping the bootstrap compiler)?
+
+Will updates/bugfixes to the code require further workarounds to CTFE
+limitations?
+
+Are the changes making a decision for many parts of the standard library?  For
+example, if a trait is made const, that likely means all implementations in the
+standard library have to be const too. It may not be the case that this is easy
```

Comment:
This is never and will never be true. A const trait can have non-const impls, otherwise we'd not be able to mark any traits as const.

Reply:
Okay, maybe a poor example. I suspect there are some things like this though? e.g., my understanding is that today a const impl is either all const (for every method). I'm happy to reword with help on where the limitation actually is.

(Obviously future work may remove those constraints, but this document is very much emphasizing *current* reality - or even cfg(bootstrap) reality).

Reply:
ah, yes, that makes a lot more sense ^^. Either all default methods on a trait are const or none. 

While I disagree with the policy (we could just make them all const for now and revert the constness if it becomes a problem), I'm fine following it to save everyone headaches